### ISSUE TITLE
Use Docker instead of regular installer

### ISSUE THREAD

**Issue by tananaev** [traccar/traccar#4048]
I have been thinking of ways to migrate to newer versions of Java and simplifying/unifying the whole installation and packaging of Traccar. Docker seems like a good modern option. We won't need a lot of dependencies that we have now. We won't need YAJSW, inno setup and any of the other packaging. A lot of those tools and outdated and poorly maintained at this point. We also will be able to use latest Java version. There won't be any platform-specific issues.

The drawback is that users would have to install docker, but I think it's pretty straight-forward process and most people should manage.

@Abyss777, what do you think?

**Comment by Abyss777** [comment]
I would suggest check this post before https://thehftguy.com/2016/11/01/docker-in-production-an-history-of-failure/
(In Russian https://habr.com/post/332450/)

My opinion: I do not understand this hype related to docker. As for me as system administrator, I do not want to install some intermediate software (with some problems) when I can do not do it.

I understand you, you have to spend much time preparing release. But I believe with docker problems will be, at least answering tons of questions "how to ..."

Have you check another java open projects, how they package releases? Like Openfire https://www.igniterealtime.org/projects/openfire/

**Comment by tananaev** [comment]
But you still need to install Java currently. With docker you won't need to do that. So, you still install a single third party package. Docker is open source, as far as I know, so proprietary software is not really an issue. Also, it's popular enough so we can find fixes for any potential problems.

As for Openfire, they seem to use Install4J. I tried it before and it's mostly designed for desktop applications. It doesn't work very well for services.

I'm also thinking of providing pure Java package as we do now ("other" option), so anyone who doesn't want to use Docker can use it directly.

**Comment by Abyss777** [comment]
> But you still need to install Java currently.

`sudo apt-get install openjdk-8-jre`

And it will be automatically updated with other distro packages. Will it with docker?
Any way Linux users should cope with any stuff... :)

But what about Windows?

**Comment by tananaev** [comment]
Installing Docker is also a one-liner on Linux, as far as I know. I'm open to any other suggestions on how we can make it simple for everyone.

Maybe we should just make native Linux packages? Then Java would be a simple package dependency. But I'm not sure how hard it would be to maintain for all various Linux packaging systems. Also, Windows would still be a problem in that case.

**Comment by Abyss777** [comment]
What about bundle java in package? I believe it is frequent approach for java applications.
Just googled http://launch4j.sourceforge.net/ can it help?

**Comment by tananaev** [comment]
I have researched it a while ago and none of those bundlers work very well. Some of them only support one platform (e.g. Windows). Others are not even designed for services.

Another problem is YAJSW or any alternative. It's very poorly supported and I think it will die very soon. It already has various compatibility issues with newer versions of Java.

**Comment by Abyss777** [comment]
Yes, "servicing" is the problem. Maybe docker is reasonable then. But vote for "pure java package" in addition :)
Is the migration for users will be a challenge?

**Comment by tananaev** [comment]
I'm sure there will be some challenges. That's why I wanted to discuss it first. If we go ahead with it, we need to make migration as easy as possible.

**Comment by Abyss777** [comment]
I think it is possible for some time for current installations make update using "pure java package" until we polish migration to docker.

**Comment by hallmark9** [comment]
Vote +1 for keeping the "pure java package" option :)

**Comment by hallmark9** [comment]
Adding a note about YAJSW just in case it's not been observed yet, there's a fork of YAJSW with active development and a stable release for newer JDK:
https://github.com/yajsw/yajsw

**Comment by tananaev** [comment]
@Abyss777, I decided to give native service another chance. I'm thinking having just two versions:

1. Linux - just use systemd to run Traccar as a service
2. Windows - implement service callbacks directly in Traccar similar to how YAJSW does it

What do you think? Are you familiar with systemd?

**Comment by Abyss777** [comment]
I like it. Native always better. I'm not very familiar with systemd, but I think it is not very complicated and well documented.
The useful thing in YAJSW was handling wrapper.log, we will have to do it our self. 

Note for future: https://stackoverflow.com/questions/21503883/spring-boot-application-as-a-service/22121547#22121547

**Comment by Abyss777** [comment]
Hm, looks like it is not a problem at all 
```
StandardOutput=/var/log1.log
StandardError=/var/log2.log
```

**Comment by tananaev** [comment]
I have moved all logging into a single file. Check my commit from earlier today.

**Comment by tananaev** [comment]
Haven't checked SMPP, but everything else seems to work.

**Comment by Abyss777** [comment]
It works
```
2018-09-14 09:38:05  INFO: Started o.e.j.s.ServletContextHandler@2b170932{/api,null,AVAILABLE}
2018-09-14 09:38:05  INFO: Started o.e.j.s.ServletContextHandler@28bf82d9{/console,null,AVAILABLE}
2018-09-14 09:38:05  INFO: Started o.t.w.@1c7f6e96{/,null,AVAILABLE}
2018-09-14 09:38:05  INFO: Started ServerConnector@d969452{HTTP/1.1}{0.0.0.0:8082}
2018-09-14 09:38:05  INFO: Started @7498ms
2018-09-14 09:38:14  INFO: sync send PDU: (bind_transceiver: 0x00000023 0x00000009 0x00000000 0x00000001) (body: systemId [user] password [password] systemType [] interfaceVersion [0x34] addressRange (0x00 0x00 [])) (opts: )
2018-09-14 09:38:15  INFO: received PDU: (bind_transceiver_resp: 0x00000017 0x80000009 0x00000000 0x00000001 result: "OK") (body: systemId [rgsmpp]) (opts: )
2018-09-14 09:38:15  INFO: SMPP session connected
2018-09-14 09:39:15  INFO: sync send PDU: (enquire_link: 0x00000010 0x00000015 0x00000000 0x00000002) (body: ) (opts: )
2018-09-14 09:39:15  INFO: received PDU: (enquire_link_resp: 0x00000010 0x80000015 0x00000000 0x00000002 result: "OK") (body: ) (opts: )
```

**Comment by tananaev** [comment]
Perfect. I have also just added support for Windows service (5d819b2776328e914a26ae011692084e1b614e26). Haven't tested it yet though. It allows to install/uninstall and run our Java process as Windows service.

**Comment by tananaev** [comment]
I have implemented both Windows and Linux services. Here are preview versions of Traccar:

- [Traccar 4.1 preview for Linux](https://www.dropbox.com/s/mq9a270nf4tp9qi/traccar-linux-4.1-preview.zip?dl=0)
- [Traccar 4.1 preview for Windows](https://www.dropbox.com/s/ngysw4q3ckswhs4/traccar-windows-4.1-preview.zip?dl=0)

Windows has been tested on Windows 10. Linux version has been tested on recent Ubuntu release. If you have some time, please test it as well. Maybe I have missed something.

**Comment by tananaev** [comment]
Second preview with Java integrated:

- [Traccar 4.1 preview 2 for Linux](https://www.dropbox.com/s/y4y2qpua40632t4/traccar-linux-64-4.1-preview2.zip?dl=0)
- [Traccar 4.1 preview 2 for Windows](https://www.dropbox.com/s/4rlypji77habggm/traccar-windows-64-4.1-preview2.zip?dl=0)

Now Traccar works out of the box without any special requirements.

I have also upgraded code to use Java 8, so we can now use all latest Java features as soon as we want to.

Closing this issue as we can live without Docker for now.

**Comment by gurabli** [comment]
Just my two cents, as a new user of Traccar: Docker is the very best thing that happened recently (after it became widely used). I am fully comfortable with managed my Linux servers without it, but since I learned the Docker basics, there is no turning back. It makes a relatively complicated system a joy to manage, backup, test, change, update, downgrade, configure. I can build a fully running complete media server with special iptables, selective VPN tunnel, VPN server, B2 backup, etc, etc in no time, with very little interaction. And the underlying system is a minimal one with only security updates. I really can't imagine anyomore to maintain something on non-Docker env. What was many hours before, now is basically a good written version of docker-compose file and teo scripts (one for installing Docker and second for setting some env variables to make docker-compose even faster and easier to use). One thing I must say: don't have any idea about Windows, as once I turned to Linux, I never looked back to that bunch of limited and crappy piece of code. For everyday use it does the job, but running a server like process? It is like if you were your own enemy.
